Add mmlu

tusharkhot opened the pull request:
Added MMLU task to catwalk
- All 57 tasks are added but have to be specified individually. Ideally we would have a Catwalk "GroupedTask" such that by only specifying `--task mmlu`, all the 57 tasks could be run. This would also allow us to aggregate the statistics across the subtasks. This seems beyond my ability but let me know if there is a simple fix.
- I am using the top-k dev examples as few-shot. The easiest way was to define a special `EleutherMMLUTask` for this
- I am using the `acc_raw` metric but I am not sure if that is the ideal one for MMLU.

OyvindTafjord commented:
This will presumably not keep the initial instruction in the usual MMLU prompt (including the updated Eleuther one: https://github.com/EleutherAI/lm-evaluation-harness/blob/master/lm_eval/tasks/hendrycks_test.py#L134). There's a few way to incorporate that, each being a bit clunky (e.g., add `prefix` option to `instance_as_rank_classification` which is then filled by some appropriate method in the task). There's also a question whether that instruction should be there in the 0-shot case (the Eleuther code doesn't include it I think, which feels like a bug).

Doing the aggregation over subjects within catwalk would be nice, will require some thought as to how to do this (and how to format the output, as we'd want to also keep the per-subject and per-instance data).

One minor thing I suggest about the prompt, changing it slightly from the Eleuther one, is to add a space before each answer choice: 
```
            choices = "".join(
                [f" {key}. {choice}\n" for key, choice in zip(keys, doc["choices"])]
            )
```
for consistent tokenization in the prompt and in the answer (e.g., " A" both places, vs "A" and " A")

tusharkhot commented:
I was thinking whether we should make that change (to add spacing before choices) in https://github.com/allenai/catwalk/blob/olmo-eval/catwalk/tasks/eleuther.py#L159 for all tasks?

tusharkhot commented:
At first I thought that we should make that change for all ranked classification tasks by overwriting the `_process_doc` function?

tusharkhot commented:
Struggling to find where the context creation is happening for RankedClassification tasks. Any pointers?

OyvindTafjord commented:
The call to `instance_as_rank_classification` happens here (through slight indirection) for the models we care about: https://github.com/allenai/catwalk/blob/olmo-eval/catwalk/models/language_model.py#L126. 

Rather than messing with that, possibly the `instance_as_rank_classification` method can check if the inner task has some sort of "description" set and initialize `prefix` with it (might need to be added as a separate field to the task class, as pulling it out from, say, https://github.com/EleutherAI/lm-evaluation-harness/blob/master/lm_eval/tasks/hendrycks_test.py#L134 isn't very easy)

Btw, the space I'm talking about is in the middle of the instance prompt, so no central way to change that.

tusharkhot commented:
I did discover that call but I dont know how the context is being created for each instance. I dug a bit more and here is my current understanding.
- The context is created in run_loglikeihood using the first entry of the tuple constructed here:
https://github.com/allenai/catwalk/blob/olmo-eval/catwalk/models/language_model.py#L140C5-L140C5
- These instance_requests are being created inside the `choices` field of RankedClassificationInstance here (where the prefix field is used)
https://github.com/allenai/catwalk/blob/olmo-eval/catwalk/tasks/eleuther.py#L161-L165
- The requests are being constructed here using the inner task function 
  https://github.com/allenai/catwalk/blob/olmo-eval/catwalk/tasks/eleuther.py#L111-L113
 
 So it should include the initial instruction in the MMLU prompt. And I think it does. This is the first field in `model_input` tuple:
 
 ```
 The following are multiple choice questions (with answers) about abstract algebra.

The cyclic subgroup of Z_24 generated by 18 has order
A. 4
B. 8
C. 12
D. 6
Answer: A

The following are multiple choice questions (with answers) about abstract algebra.

Find the order of the factor group Z_6/<3>.
A. 2
B. 3
C. 6
D. 12
Answer: B

The following are multiple choice questions (with answers) about abstract algebra.

Statement 1 | A permutation that is a product of m even permutations and n odd permutations is an even permutation if and only if n is even. Statement 2 | Every group is isomorphic to a group of permutations.
A. True, True
B. False, False
C. True, False
D. False, True
Answer: A

The following are multiple choice questions (with answers) about abstract algebra.

The cyclic subgroup of Z_24 generated by 18 has order
A. 4
B. 8
C. 12
D. 6
Answer:  A
```

Maybe it should not repeat for every instance?

OyvindTafjord commented:
Right, we don't want that instruction repeated, so might need to do some surgery on the inner task to avoid this which is a bit annoying.

tusharkhot commented:
As per our discussion, Changed the top-level function for generating ranked classification instances. Now we get correctly formatted instances such as:

```
The following are multiple choice questions (with answers) about abstract algebra.

Find all c in Z_3 such that Z_3[x]/(x^2 + c) is a field.
 A. 0
 B. 1
 C. 2
 D. 3
Answer: B

Statement 1 | If aH is an element of a factor group, then |aH| divides |a|. Statement 2 | If H and K are subgroups of G then HK is a subgroup of G.
 A. True, True
 B. False, False
 C. True, False
 D. False, True
Answer: B

Statement 1 | Every element of a group generates a cyclic subgroup of the group. Statement 2 | The symmetric group S_10 has 10 elements.
 A. True, True
 B. False, False
 C. True, False
 D. False, True
Answer: C

The cyclic subgroup of Z_24 generated by 18 has order
 A. 4
 B. 8
 C. 12
 D. 6
Answer:
```
Also added the space before choices

OyvindTafjord reviewed (approved):
Thanks, LGTM! Maybe put in a comment in the hendrycks_test.py file that the prefix spaces was added (and is thus slightly different from the original Eleuther task).

tusharkhot commented:
Good pt. Done!
